### PR TITLE
[FEA] Add dot product binary op

### LINKED ISSUE #2448: [FEA] Dot Product
Seems like there's no easy way for dot product can this be implemented?

Pandas equivalent: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dot.html

### PR DESCRIPTION
**Pull request by charlesbluca**
Closes #2448 

Implements dot product for dataframes and series by using `cupy.ndarray.dot` on their underlying `values` property.

### CODE REVIEW

File: python/cudf/cudf/core/frame.py [hunk 1]
```diff
@@ -3594,6 +3594,64 @@ def _colwise_binop(
...
+            0    1
+        0  30   70
+        1  70  174
+        >>> s = cudf.Series([1, 1, 1, 1])
+        >>> df @ s
+        0    10
+        1    26
+        dtype: int64
+        >>> [1, 2, 3, 4] @ s
+        array(10)
```
Comment (charlesbluca) [inline python/cudf/cudf/core/frame.py:3633]: Taking the dot product of two vectors (1-dimensional cupy arrays) gives us a 0-d array - do we want to extract the value from this, and if so, is there a preferred way to do this?

File: python/cudf/cudf/core/frame.py [hunk 2]
```diff
@@ -3594,6 +3594,64 @@ def _colwise_binop(
...
+            return NotImplemented
+        if reflect:
+            lhs, rhs = rhs.T, lhs.T
+
+        result = lhs.dot(rhs)
+        if len(result.shape) == 1:
+            result = cudf.Series(result)
+        elif len(result.shape) == 2:
+            result = cudf.DataFrame(result)
+
```
Comment (skirui-source) [inline python/cudf/cudf/core/frame.py:3652]: What about for case when returned value is  a scalar (single value)?

File: python/cudf/cudf/core/frame.py (hunk 1 above)
Comment (skirui-source) [reply python/cudf/cudf/core/frame.py:3633]: Looks like you can either convert to a specific type, eg int(x) or float(x) or  using x.item() where x =the 0-d array.
source: https://stackoverflow.com/questions/9814226/error-extracting-element-from-an-array-python

File: python/cudf/cudf/core/frame.py [hunk 3]
```diff
@@ -3594,6 +3594,64 @@ def _colwise_binop(
...
+        Among flexible wrappers (`add`, `sub`, `mul`, `div`, `mod`, `pow`,
+        `dot`) to arithmetic operators: `+`, `-`, `*`, `/`, `//`, `%`, `**`,
+        `@`.
+
+        Parameters
+        ----------
+        other : sequence, Series, or DataFrame
+            Any multiple element data structure, or list-like object.
+        reflect : bool, default False
+            If ``True`` the operation is reflected (i.e whether to swap the
```
[the remark this answers is not in the document]
Comment (charlesbluca) [reply python/cudf/cudf/core/frame.py:3610]: In the context of the binary ops, this indicates that `other` is on the left hand side instead of the right hand side; in cases where we are doing dot products using the `@` binary operator, this would be necessary if our LHS doesn't implement `__matmul__()`. Essentially

```python
cudf_frame_object @ [1, 2, 3, 4]  # calls Frame.__matmul__() -> Frame.dot(...reflect=False)
[1, 2, 3, 4] @ cudf_frame_object  # calls Frame.__rmatmul__() -> Frame.dot(...reflect=True)
```

It would probably be helpful to clarify this a bit better in the docstrings.

File: python/cudf/cudf/core/frame.py (hunk 2 above)
Comment (charlesbluca) [reply python/cudf/cudf/core/frame.py:3652]: Based on your other review comment, we should probably do something like

```python
result = result.item()
```

File: python/cudf/cudf/core/frame.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (skirui-source) [reply python/cudf/cudf/core/frame.py:3610]: thanks for the explanation, yes i also think you should clarify that further in the docstrings :))

File: python/cudf/cudf/tests/test_binops.py [hunk 4]
```diff
@@ -2913,3 +2913,24 @@ def test_empty_column(binop, data, scalar):
     expected = binop(pdf, scalar)
 
     utils.assert_eq(expected, got)
+
+
+@pytest.mark.parametrize(
+    "lhs",
+    [pd.DataFrame([[0, 1, -2, -1], [1, 1, 1, 1]]), pd.Series([1, 1, 2, 1])],
```
[the remark this answers is not in the document]
Comment (charlesbluca) [reply python/cudf/cudf/tests/test_binops.py:2920]: Right now null value cases will have to be xfailed, as we cannot get the `values` attribute of a series/dataframe with nulls.

**Comment by charlesbluca** [comment]
rerun tests

**Review by skirui-source — APPROVED** [review]

**Review by marlenezw — APPROVED** [review]
Changes look good to me!

**Comment by charlesbluca** [comment]
rerun tests

File: python/cudf/cudf/tests/test_binops.py [hunk 5]
```diff
@@ -2913,3 +2913,52 @@ def test_empty_column(binop, data, scalar):
...
+        pd.DataFrame([[9.4, 10.5], [11.6, 12.7], [13.8, 14.9], [15.1, 16.2]]),
+        pd.Series([5, 6, 7, 8]),
+        pd.Series([5.6, 6.7, 7.8, 8.9]),
+    ],
+)
+def test_binops_dot(lhs, rhs):
+    if isinstance(lhs, list):
+        glhs = lhs
+    else:
+        glhs = cudf.from_pandas(lhs)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_binops.py:2947]:
```suggestion
    glhs = lhs if isinstance(lhs, list) else cudf.from_pandas(lhs)
```
Should we be supporting dot between a pandas object and a cudf object? If so, perhaps you should be injecting this logic into `dot` rather than here. Same goes for `rhs` below (of course, at least one operand must be a `cudf` object to trigger our logic).

File: python/cudf/cudf/tests/test_binops.py [hunk 6]
```diff
@@ -2913,3 +2913,52 @@ def test_empty_column(binop, data, scalar):
...
+
+    if isinstance(rhs, list):
+        grhs = rhs
+    else:
+        grhs = cudf.from_pandas(rhs)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_binops.py:2952]:
```suggestion
    grhs = rhs if isinstance(rhs, list) else cudf.from_pandas(rhs)
```

File: python/cudf/cudf/tests/test_array_function.py [hunk 7]
```diff
@@ -60,7 +60,6 @@ def test_array_func_cudf_dataframe(pd_df, func):
     "func",
     [
         lambda x: np.cov(x, x),
-        lambda x: np.dot(x, x),
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_array_function.py:63]: Why was this removed? We still want the dispatch to work, right?

File: python/cudf/cudf/core/frame.py [hunk 8]
```diff
@@ -3594,6 +3594,67 @@ def _colwise_binop(
...
+        dtype: int64
+        >>> [1, 2, 3, 4] @ s
+        10
+        """
+        lhs = self.values
+        if isinstance(other, Frame):
+            rhs = other.values
+        elif isinstance(other, cupy.ndarray):
+            rhs = other
+        elif isinstance(other, abc.Sequence):
```
Comment (vyasr) [inline python/cudf/cudf/core/frame.py:3641]:
```suggestion
        elif isinstance(other, (abc.Sequence, np.ndarray)):
```

File: python/cudf/cudf/core/frame.py [hunk 9]
```diff
@@ -3594,6 +3594,67 @@ def _colwise_binop(
...
+
+        result = lhs.dot(rhs)
+        if len(result.shape) == 1:
+            result = cudf.Series(result)
+        elif len(result.shape) == 2:
+            result = cudf.DataFrame(result)
+        else:
+            result = result.item()
+
+        return result
```
Comment (vyasr) [inline python/cudf/cudf/core/frame.py:3656]: This is a bit more idiomatic in Python and results in returning as fast as possible. Not too important here, but a good practice to follow.
```suggestion
        result = lhs.dot(rhs)
        if len(result.shape) == 1:
            return cudf.Series(result)
        if len(result.shape) == 2:
            return cudf.DataFrame(result)
        return result.item()
```

File: python/cudf/cudf/core/frame.py [hunk 10]
```diff
@@ -3594,6 +3594,67 @@ def _colwise_binop(
...
+        """
+        Get dot product of frame and other, (binary operator `dot`).
+
+        Among flexible wrappers (`add`, `sub`, `mul`, `div`, `mod`, `pow`,
+        `dot`) to arithmetic operators: `+`, `-`, `*`, `/`, `//`, `%`, `**`,
+        `@`.
+
+        Parameters
+        ----------
+        other : sequence, Series, or DataFrame
```
Comment (vyasr) [inline python/cudf/cudf/core/frame.py:3607]:
```suggestion
        other : Sequence, Series, or DataFrame
```

**Review by vyasr — CHANGES_REQUESTED** [review]
This looks fine overall. It's unfortunate that it has to bypass the other binop machinery, but since the dot product is not strictly column by column I don't see a good alternative. Also we're dispatching to cupy rather quickly so it's not a huge issue.

File: python/cudf/cudf/tests/test_array_function.py (hunk 7 above)
Comment (charlesbluca) [reply python/cudf/cudf/tests/test_array_function.py:63]: I thought this test was assuming that DataFrames don't have `dot()`, and thus is checking for a `TypeError` when we try to call the lambda function on it (at least that's what I gathered from the initial test failures) - is that not the case?

File: python/cudf/cudf/tests/test_binops.py (hunk 5 above)
Comment (charlesbluca) [reply python/cudf/cudf/tests/test_binops.py:2947]: You're right, we should be supporting dot between pandas and cuDF - I'll add that in and make some tests using pandas dataframes.

File: python/cudf/cudf/tests/test_array_function.py (hunk 7 above)
Comment (vyasr) [reply python/cudf/cudf/tests/test_array_function.py:63]: Oh you are 100% right, my bad. I just didn't check the body of the test. We should probably add suitable tests for this, then. It looks like previously `np.dot(cudf.Series([1, 2, 3]), cudf.Series([1, 2, 3]))` would work by internal dispatch to `cupy.dot`. Does it now go through `cudf.Frame.dot`, and it works for both `DataFrame` and `Series`? That should get tested now I think.

Comment (charlesbluca) [reply python/cudf/cudf/tests/test_array_function.py:63]: It looks like `np.dot()` is something of a special case for Pandas, in that it returns a Numpy array instead of a DataFrame/Series - is this something we should seek to replicate in cuDF's `__array_function__()`s, or just make a workaround for in the tests (i.e. if we are using `np.dot()` compare the `values` of the result)?

Comment (vyasr) [reply python/cudf/cudf/tests/test_array_function.py:63]: I think we should probably aim for the most equivalent approach that also won't destroy performance. In this case, that probably means conversion to a `cupy` array by inserting something a call to the `.values` property in the return statements.

**Comment by vyasr** [comment]
@charlesbluca is this ready for another round of review?

**Comment by charlesbluca** [comment]
Yup, added a special case in `__array_function__` to return the proper output.

**Review by vyasr — APPROVED** [review]
Looks good! I have a couple of higher-level questions, but nothing that needs to hold this up.

1. I noticed that you added some more descriptive comments for reflected ops. Was that requested during review? Perhaps we should instead include links to the Python Data Model page, specifically [the section that describes reflected binops](https://docs.python.org/3/reference/datamodel.html#object.__ror__).
2. I get that we try to match pandas everywhere, including docs, but maintaining N copies of comments like the "Among flexible wrappers..." seems bound to get out of date. Do we really need that? If we do, perhaps we should use a formatted template for those docstrings to reduce repetition (although that would be out of scope for this PR).

**Comment by charlesbluca** [comment]
Thanks @vyasr!

1. Yeah, I did that based on a review thread with @skirui-source [here](https://github.com/rapidsai/cudf/pull/8909#discussion_r689854235); thanks for that link, it definitely explains the concept in greater detail and can be used in place of my explanation.
2. Agreed, maintaining those docstrings is a lot of extra toil that can be avoided - are there docstrings in the codebase using templates currently?

**Comment by vyasr** [comment]
Regarding templates: yes, but very haphazardly and certainly not with any consistent plan across the code base. I opened #9050 to discuss this further.

Feel free to merge this whenever you've updated the documentation on reflection as you see fit.

**Comment by charlesbluca** [comment]
rerun tests

**Comment by charlesbluca** [comment]
@gpucibot merge
